File: cmd/secureyeoman-edge/sandbox.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

// SandboxConfig controls which commands may run and how output is bounded.
type SandboxConfig struct {
	WorkspaceDir    string   `json:"workspaceDir"`    // root directory for file access, default "."
	AllowedCommands []string `json:"allowedCommands"` // empty means allow all (subject to blocklist)
	BlockedCommands []string `json:"blockedCommands"` // always blocked regardless of allowlist
	MaxOutputBytes  int      `json:"maxOutputBytes"`  // default 1 MB
	TimeoutSeconds  int      `json:"timeoutSeconds"`  // default 30
	Enabled         bool     `json:"enabled"`
}

// ExecResult holds the outcome of a sandboxed command execution.
type ExecResult struct {
	Command   string `json:"command"`
	ExitCode  int    `json:"exitCode"`
	Stdout    string `json:"stdout"`
	Stderr    string `json:"stderr"`
	Duration  int64  `json:"durationMs"`
	Truncated bool   `json:"truncated,omitempty"`
}

// Sandbox enforces command allowlists and workspace path scoping.
type Sandbox struct {
	config SandboxConfig
	logger *Logger
}

var defaultAllowedCommands = []string{
	"ls", "cat", "head", "tail", "wc", "grep", "find",
	"df", "du", "uname", "hostname", "whoami", "date", "uptime",
	"ping", "curl", "wget", "ip", "ss", "ps", "top", "free",
	"lsblk", "lscpu", "sensors", "journalctl",
}

var defaultBlockedCommands = []string{
	"rm", "dd", "mkfs", "shutdown", "reboot", "poweroff", "halt",
	"init", "kill", "killall", "pkill", "mount", "umount",
	"fdisk", "parted", "iptables", "nft",
}

// NewSandbox creates a Sandbox with defaults applied for any zero-value fields.
func NewSandbox(config SandboxConfig, logger *Logger) *Sandbox {
	if config.WorkspaceDir == "" {
		config.WorkspaceDir = "."
	}
	if len(config.AllowedCommands) == 0 {
		config.AllowedCommands = defaultAllowedCommands
	}
	if len(config.BlockedCommands) == 0 {
		config.BlockedCommands = defaultBlockedCommands
	}
	if config.MaxOutputBytes <= 0 {
		config.MaxOutputBytes = 1 << 20 // 1 MB
	}
	if config.TimeoutSeconds <= 0 {
		config.TimeoutSeconds = 30
	}
	return &Sandbox{config: config, logger: logger}
}

// ListAllowed returns the current allowlist of command binaries.
func (s *Sandbox) ListAllowed() []string {
	out := make([]string, len(s.config.AllowedCommands))
	copy(out, s.config.AllowedCommands)
	return out
}

// ValidateCommand checks whether a command string is permitted to run.
// It returns a non-nil error if the command is blocked, not on the allowlist,
// or contains path traversal outside the workspace.
func (s *Sandbox) ValidateCommand(command string) error {
	fields := strings.Fields(command)
	if len(fields) == 0 {
		return fmt.Errorf("empty command")
	}

	binary := filepath.Base(fields[0])

	// Check blocklist first — these are always denied.
	for _, blocked := range s.config.BlockedCommands {
		if binary == blocked {
			return fmt.Errorf("command %q is blocked", binary)
		}
	}

	// Check allowlist when it is non-empty.
	if len(s.config.AllowedCommands) > 0 {
		allowed := false
		for _, a := range s.config.AllowedCommands {
			if binary == a {
				allowed = true
				break
			}
		}
		if !allowed {
			return fmt.Errorf("command %q is not in the allowlist", binary)
		}
	}

	// Reject path traversal in any argument.
	workspaceAbs, err := filepath.Abs(s.config.WorkspaceDir)
	if err != nil {
		return fmt.Errorf("cannot resolve workspace directory: %w", err)
	}
	for _, arg := range fields[1:] {
		if !s.isArgPathSafe(arg, workspaceAbs) {
			return fmt.Errorf("argument %q attempts path traversal outside workspace", arg)
		}
	}

	return nil
}

// isArgPathSafe returns false when arg looks like a path that resolves outside
// workspaceAbs. Bare flags and non-path tokens are left unrestricted.
func (s *Sandbox) isArgPathSafe(arg, workspaceAbs string) bool {
	// Only scrutinise tokens that contain path separators or ".."
	if !strings.Contains(arg, "/") && !strings.Contains(arg, "..") {
		return true
	}
	if strings.Contains(arg, "..") {
		cleaned := filepath.Clean(arg)
		if filepath.IsAbs(cleaned) {
			// Absolute path — must be inside workspace.
			return strings.HasPrefix(cleaned, workspaceAbs+string(filepath.Separator)) ||
				cleaned == workspaceAbs
		}
		// Relative path with ".." — join with workspace and verify.
		joined := filepath.Clean(filepath.Join(workspaceAbs, cleaned))
		return strings.HasPrefix(joined, workspaceAbs+string(filepath.Separator)) ||
			joined == workspaceAbs
	}
	return true
}

// IsPathAllowed reports whether path (absolute or relative) falls within the
// configured WorkspaceDir. Symlinks are resolved before the prefix check so
// that a symlink pointing outside the workspace is correctly rejected.
func (s *Sandbox) IsPathAllowed(path string) bool {
	workspaceAbs, err := filepath.Abs(s.config.WorkspaceDir)
	if err != nil {
		return false
	}
	// Resolve symlinks on the workspace root itself.
	if resolved, err := filepath.EvalSymlinks(workspaceAbs); err == nil {
		workspaceAbs = resolved
	}

	var absPath string
	if filepath.IsAbs(path) {
		absPath = filepath.Clean(path)
	} else {
		absPath = filepath.Clean(filepath.Join(workspaceAbs, path))
	}
	// Resolve symlinks on the target path before checking the prefix.
	if resolved, err := filepath.EvalSymlinks(absPath); err == nil {
		absPath = resolved
	}
	return absPath == workspaceAbs ||
		strings.HasPrefix(absPath, workspaceAbs+string(filepath.Separator))
}

// Execute runs a shell command inside the sandbox and returns its result.
// The sandbox must be Enabled; otherwise an error is returned immediately.
func (s *Sandbox) Execute(command string) (*ExecResult, error) {
	if !s.config.Enabled {
		return nil, fmt.Errorf("sandbox is disabled")
	}

	if err := s.ValidateCommand(command); err != nil {
		return nil, fmt.Errorf("command validation failed: %w", err)
	}

	timeout := time.Duration(s.config.TimeoutSeconds) * time.Second
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	// Split the command into binary + args for exec.CommandContext to avoid
	// shell injection. We deliberately do NOT use "sh -c" here.
	fields := strings.Fields(command)
	binary := fields[0]
	args := fields[1:]

	cmd := exec.CommandContext(ctx, binary, args...)
	cmd.Dir = s.config.WorkspaceDir

	// Inherit the current environment but override HOME to the workspace root.
	env := inheritEnvWithOverride("HOME", s.config.WorkspaceDir)
	cmd.Env = env

	var stdoutBuf, stderrBuf bytes.Buffer
	cmd.Stdout = &stdoutBuf
	cmd.Stderr = &stderrBuf

	start := time.Now()
	runErr := cmd.Run()
	durationMs := time.Since(start).Milliseconds()

	exitCode := 0
	if runErr != nil {
		if exitErr, ok := runErr.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else if ctx.Err() == context.DeadlineExceeded {
			exitCode = -1
		} else {
			return nil, fmt.Errorf("exec error: %w", runErr)
		}
	}

	stdoutBytes := stdoutBuf.Bytes()
	stderrBytes := stderrBuf.Bytes()
	truncated := false

	if len(stdoutBytes)+len(stderrBytes) > s.config.MaxOutputBytes {
		truncated = true
		remaining := s.config.MaxOutputBytes
		if len(stdoutBytes) > remaining {
			stdoutBytes = stdoutBytes[:remaining]
			remaining = 0
		} else {
			remaining -= len(stdoutBytes)
		}
		if len(stderrBytes) > remaining {
			stderrBytes = stderrBytes[:remaining]
		}
	}

	result := &ExecResult{
		Command:   command,
		ExitCode:  exitCode,
		Stdout:    string(stdoutBytes),
		Stderr:    string(stderrBytes),
		Duration:  durationMs,
		Truncated: truncated,
	}

	s.logger.Debug("sandbox exec",
		"command", command,
		"exitCode", exitCode,
		"durationMs", durationMs,
		"truncated", truncated,
	)

	return result, nil
}

// inheritEnvWithOverride returns os.Environ() with key set to value.
func inheritEnvWithOverride(key, value string) []string {
	prefix := key + "="
	base := make([]string, 0, 32)
	for _, e := range os.Environ() {
		if !strings.HasPrefix(e, prefix) {
			base = append(base, e)
		}
	}
	base = append(base, prefix+value)
	return base
}

File: crates/sy-edge/src/sandbox.rs
//! Sandbox — allowlist-based command execution with workspace scoping.

use std::process::Command;

#[derive(Debug)]
pub struct ExecOutput {
    pub stdout: String,
    pub stderr: String,
    pub exit_code: i32,
}

const DEFAULT_ALLOWED: &[&str] = &[
    "ls", "cat", "head", "tail", "wc", "grep", "find", "df", "du", "uname", "hostname", "ping",
    "curl", "wget", "ip", "ss", "ps", "top", "free", "lsblk", "lscpu", "sensors", "journalctl",
];

const BLOCKED: &[&str] = &[
    "rm", "dd", "mkfs", "shutdown", "reboot", "poweroff", "halt", "init", "kill", "pkill",
    "mount", "fdisk", "iptables", "nft",
];

const MAX_OUTPUT: usize = 1_048_576; // 1 MB

pub struct SandboxManager {
    allowed: Vec<String>,
}

impl SandboxManager {
    pub fn new() -> Self {
        Self {
            allowed: DEFAULT_ALLOWED.iter().map(|s| s.to_string()).collect(),
        }
    }

    pub fn execute(
        &self,
        command: &str,
        args: &[String],
        workspace: Option<&str>,
        timeout_secs: u64,
    ) -> Result<ExecOutput, String> {
        // Check against blocklist
        let cmd_base = command.split('/').last().unwrap_or(command);
        if BLOCKED.contains(&cmd_base) {
            return Err(format!("Command blocked: {cmd_base}"));
        }

        // Check allowlist
        if !self.allowed.iter().any(|a| a == cmd_base) {
            return Err(format!("Command not allowed: {cmd_base}"));
        }

        // Validate workspace path (prevent traversal)
        if let Some(ws) = workspace {
            let canonical = std::fs::canonicalize(ws)
                .map_err(|e| format!("Invalid workspace: {e}"))?;
            if !canonical.starts_with("/tmp") && !canonical.starts_with("/home") {
                return Err("Workspace must be under /tmp or /home".into());
            }
        }

        // Validate args don't contain path traversal
        for arg in args {
            if arg.contains("..") {
                return Err("Path traversal detected in arguments".into());
            }
        }

        let mut cmd = Command::new(command);
        cmd.args(args);
        if let Some(ws) = workspace {
            cmd.current_dir(ws);
        }

        let output = cmd
            .output()
            .map_err(|e| format!("Execution failed: {e}"))?;

        // Wait with timeout (basic — Command::output blocks)
        let _ = timeout_secs; // TODO: implement proper timeout with tokio

        let mut stdout = String::from_utf8_lossy(&output.stdout).to_string();
        let mut stderr = String::from_utf8_lossy(&output.stderr).to_string();

        // Truncate to max output
        stdout.truncate(MAX_OUTPUT);
        stderr.truncate(MAX_OUTPUT);

        Ok(ExecOutput {
            stdout,
            stderr,
            exit_code: output.status.code().unwrap_or(-1),
        })
    }

    pub fn allowed_commands(&self) -> Vec<String> {
        self.allowed.clone()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn allowed_commands_populated() {
        let sm = SandboxManager::new();
        let cmds = sm.allowed_commands();
        assert!(cmds.contains(&"ls".to_string()));
        assert!(cmds.contains(&"cat".to_string()));
        assert!(cmds.contains(&"grep".to_string()));
        assert!(cmds.len() >= 20);
    }

    #[test]
    fn blocked_command_rejected() {
        let sm = SandboxManager::new();
        let result = sm.execute("rm", &[], None, 30);
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("blocked"));
    }

    #[test]
    fn unlisted_command_rejected() {
        let sm = SandboxManager::new();
        let result = sm.execute("python3", &[], None, 30);
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("not allowed"));
    }

    #[test]
    fn path_traversal_in_args_rejected() {
        let sm = SandboxManager::new();
        let result = sm.execute("ls", &["../../etc/passwd".to_string()], None, 30);
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("traversal"));
    }

    #[test]
    fn allowed_command_executes() {
        let sm = SandboxManager::new();
        let result = sm.execute("uname", &["-s".to_string()], None, 30);
        assert!(result.is_ok());
        let output = result.unwrap();
        assert!(output.stdout.contains("Linux"));
        assert_eq!(output.exit_code, 0);
    }

    #[test]
    fn ls_with_workspace() {
        let sm = SandboxManager::new();
        let result = sm.execute("ls", &[], Some("/tmp"), 30);
        assert!(result.is_ok());
        assert_eq!(result.unwrap().exit_code, 0);
    }

    #[test]
    fn bad_workspace_rejected() {
        let sm = SandboxManager::new();
        let result = sm.execute("ls", &[], Some("/etc"), 30);
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("Workspace must be"));
    }

    #[test]
    fn command_with_full_path_uses_basename() {
        let sm = SandboxManager::new();
        // /usr/bin/rm has basename "rm" which is blocked
        let result = sm.execute("/usr/bin/rm", &[], None, 30);
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("blocked"));
    }

    #[test]
    fn blocked_commands_comprehensive() {
        let sm = SandboxManager::new();
        for cmd in ["dd", "mkfs", "shutdown", "reboot", "kill", "mount", "iptables"] {
            let result = sm.execute(cmd, &[], None, 30);
            assert!(result.is_err(), "{cmd} should be blocked");
        }
    }

    #[test]
    fn stderr_captured() {
        let sm = SandboxManager::new();
        // ls a nonexistent path should produce stderr
        let result = sm.execute("ls", &["/nonexistent_path_xyz".to_string()], None, 30);
        assert!(result.is_ok());
        let output = result.unwrap();
        assert!(!output.stderr.is_empty());
        assert_ne!(output.exit_code, 0);
    }
}
